scheduling/taskmetadata.go

```diff
@@ -0,0 +1,533 @@
+package scheduling
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/featureform/fferr"
+	"github.com/featureform/locker"
+	ss "github.com/featureform/storage"
+)
+
+type TaskMetadataList []TaskMetadata
+
+type TaskRunList []TaskRunMetadata
+
+func (trl *TaskRunList) FilterByStatus(status Status) {
+	var newList TaskRunList
+	for _, run := range *trl {
+		if run.Status == status {
+			newList = append(newList, run)
+		}
+	}
+	*trl = newList
+}
+
+type TaskMetadataManager struct {
+	storage ss.MetadataStorage
+}
+
+func NewMemoryTaskMetadataManager() TaskMetadataManager {
+	memoryLocker := locker.MemoryLocker{
+		LockedItems: sync.Map{},
+		Mutex:       &sync.Mutex{},
+	}
+
+	memoryStorage := ss.MemoryStorageImplementation{
+		Storage: make(map[string]string),
+	}
+
+	memoryMetadataStorage := ss.MetadataStorage{
+		Locker:  &memoryLocker,
+		Storage: &memoryStorage,
+	}
+	return TaskMetadataManager{
+		storage: memoryMetadataStorage,
+	}
+}
+
+func (m *TaskMetadataManager) CreateTask(name string, tType TaskType, target TaskTarget) (TaskMetadata, fferr.GRPCError) {
+	keys, err := m.storage.List(TaskMetadataKey{}.String())
+	if err != nil {
+		err.AddDetail("metadata type", "task metadata")
```

Comment:
how we create errors is v clunky rn 3 lines to create an error, we should have a err.WithDetails(...) similar to how a logger does it. Don't worry about it on this PR though